src/modules/mahjong/MahjongGameState/MahjongGameState.ts

```diff
@@ -0,0 +1,106 @@
+import * as PIXI from 'pixi.js';
+
+import GameState from '../../game/GameState/GameState';
+import WindEnums from '../enums/WindEnums';
+import UserEntity from '../../game/UserEntity/UserEntity';
+import SpriteFactory from '../../../pixi/SpriteFactory';
+import DeadPile from '../DeadPile/DeadPile';
+import WallCounter from '../WallCounter/WallCounter';
+
+/**
+ * Class that holds all the Mahjong game state for the Front-End Client.
+ */
+class MahjongGameState extends GameState {
+  private currentWind: WindEnums;
+
+  private redrawPending: boolean;
+
+  private dealer: number;
+
+  private deadPile: DeadPile;
+
+  private wallCounter: WallCounter;
+
+  constructor(users: UserEntity[]) {
+    super(users);
+    this.currentWind = WindEnums.EAST;
+    this.redrawPending = false;
+    this.dealer = 0;
+    this.deadPile = new DeadPile();
+    this.wallCounter = new WallCounter();
+  }
+
+  /**
+   * Changes redrawPending to false to initialize a redraw
+   */
+  public requestRedraw = (): void => {
+    this.redrawPending = false;
+  };
+
+  public getCurrentWind(): WindEnums {
+    return this.currentWind;
+  }
+
+  public getDealer(): number {
+    return this.dealer;
+  }
+
+  public getDeadPile(): DeadPile {
+    return this.deadPile;
+  }
+
+  /**
+   * Changes wind (in counter clockwise direction)
+   */
+  public changeWind(): void {
+    const current = this.currentWind;
+
+    if (current === WindEnums.EAST) {
+      this.currentWind = WindEnums.SOUTH;
+    } else if (current === WindEnums.SOUTH) {
+      this.currentWind = WindEnums.WEST;
+    } else if (current === WindEnums.WEST) {
+      this.currentWind = WindEnums.NORTH;
+    } else if (current === WindEnums.NORTH) {
+      this.currentWind = WindEnums.EAST;
+    }
+  }
+
+  /**
+   * Changes dealer. If dealer goes back to 0, change wind as well.
+   */
+  public changeDealer(): void {
+    this.dealer += 1;
+    if (this.dealer >= this.getUsers().length) {
+      this.dealer = 0;
+      this.changeWind();
+    }
+  }
+
```

Comment:
To answer previous and same question, during the game (while people are playing tiles), I don't think it is necessary to constantly sync the backend (source of truth) with the front end. It is a turn based game, not a real real-time game (where we may need to constantly sync x,y positions of characters).
Perhaps we can implement syncing up game states at the end of each round (when someone wins) in case game states do get desynced by bad code or bugs.
As of now, wind, dealer are round-specific so these values can be synced up. Turn is turn specific, so there is no need to check and validate

Reply:
Thanks for the clarification. I will still implement the necessary lambda functions to sync up frontend with backend:

**Round specific**: `CHANGE_WIND`, `GET_CURRENT_WIND`, `CHANGE_DEALER`, `GET_CURRENT_DEALDER` (name can change)
**Turn specific**: `CHANGE_TURN`, `GET_CURRENT_TURN` (name can change)

and it's up to you to decide how/when you want to use it in the frontend

Reply:
In that case, will your lambda functions be calling each other if it fulfills some condition?

Reply:
Not sure if I would rely on the front end to make changes to the state if there could potentially be 4 clients sending the same message. Perhaps 1 lambda function to GET_GAME_STATE that returns everything would be better.

Reply:
GET_GAME_STATE could work, but how would you update the game state in the backend? Maybe only allow the host to send updates about the game state to the backend?

Reply:
The backend game state can update itself when some event happens, i.e. someone wins, backend checks position of the winner and updates wind/dealer accordingly.